### ISSUE TITLE
implied bounds from associated types may not actually get implied pt 2.

### ISSUE THREAD

**Issue by lcnr** [rust-lang/rust#98543]
revives #91068 which has been fixed by only considering implied bounds from projections if they don't normalize **while typechecking the function itself**

https://github.com/rust-lang/rust/blob/7125846d17a549866aa1ed0ea28778537378fca8/compiler/rustc_borrowck/src/type_check/free_region_relations.rs#L256-L289

This does **not** mean that the projection won't normalize when getting called:
```rust
trait Trait {
    type Type;
}

impl<T> Trait for T {
    type Type = ();
}

fn f<'a, 'b>(s: &'b str, _: <&'a &'b () as Trait>::Type) -> &'a str
where
    &'a &'b (): Trait, // <- adding this bound is the change from #91068 
{
    s
}

fn main() {
    let x = String::from("Hello World!");
    let y = f(&x, ());
    drop(x);
    println!("{}", y);
}
```
The added bound prevents `<&'a &'b () as Trait>::Type` from getting normalized while borrowchecking `f`, as we prefer param candidates over impl candidates. When calling `f`, we don't have the `&'a &'b (): Trait` in our `param_env`, so we can now freely use the impl candidate to normalize the projection. The caller therefore doesn't have to prove that  `&'a &'b ()` is well formed, causing unsoundness.

I am a bit surprised that the caller doesn't have to prove that the `&'a &'b (): Trait` obligation is well formed, which would cause this example to not be unsound. It doesn't remove the general unsoundness here though. Here's an alternative test where that isn't enough:
```rust
trait Trait {
    type Type;
}

impl<T> Trait for T {
    type Type = ();
}

fn f<'a, 'b>(s: &'b str, _: <&'a &'b () as Trait>::Type) -> &'a str
where
    for<'what, 'ever> &'what &'ever (): Trait,
{
    s
}

fn main() {
    let x = String::from("Hello World!");
    let y = f(&x, ());
    drop(x);
    println!("{}", y);
}
```

**Comment by steffahn** [comment]
Unlike #91068, these code examples compile “successfully” all the way since Rust 1.7.0.

**Comment by apiraino** [comment]
WG-prioritization assigning priority ([Zulip discussion](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.2398543.20implied.20bounds.20from.20associated.20types.20may.20not.20actuall.E2.80.A6)).

@rustbot label -I-prioritize +P-high

**Comment by lcnr** [comment]
while this currently does not work, we should also add a test for something like this. If we get stronger implied bounds, simply requiring the caller to prove wf for all projection components won't be enough anymore
```rust
#![feature(generic_associated_types)]

trait Trait<'a> {
    type Type
    where
        Self: 'a;
}

impl<'a, T> Trait<'a> for T {
    type Type = ()
    where
        Self: 'a; // once this bound gets implied, we have UB again
}

fn f<'a, 'b>(s: &'b str, _: <&'b () as Trait<'a>>::Type) -> &'a str
where
    &'b (): Trait<'a>, // <- adding this bound is the change from #91068 
{
    s
}

fn main() {
    let x = String::from("Hello World!");
    let y = f(&x, ());
    drop(x);
    println!("{}", y);
}
```

```rust
trait Trait<'a>: 'a {
    type Type;
}

impl<'a, T> Trait<'a> for T
where
    T: 'a // without this bound we would probably get ub here again
{
    type Type = ();
}

fn f<'a, 'b>(s: &'b str, _: <&'b () as Trait<'a>>::Type) -> &'a str
where
    &'b (): Trait<'a>, // <- adding this bound is the change from #91068 
{
    s
}

fn main() {
    let x = String::from("Hello World!");
    let y = f(&x, ());
    drop(x);
    println!("{}", y);
}
```

**Comment by jackh726** [comment]
There's a PR open that fixes this.

**Comment by lcnr** [comment]
fixed by #99217
